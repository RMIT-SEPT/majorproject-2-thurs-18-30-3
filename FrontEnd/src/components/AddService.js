import React from 'react';
import { useHistory } from "react-router-dom";
import '../containers/App.css';

// GUI for adding a service
function AddService({create}) {
    const history = useHistory();
    const[name, setName] = React.useState("");
    const[description, setDescription] = React.useState("");
    const[img, setImg] = React.useState(null);

    const handleSubmit = () => {
        if(name !== "" && description !== "")
        {
            create(name, description, img);
            history.push('/adminservices');
            window.location.reload();
        }
        else
        {
            alert("Please enter valid data");
        }
    }

    const handleImage = (e) => {
        setImg(e.target.files[0])
    }

    const validate = () => {
        if (
          name !== '' &&
          description !== '' &&
          /[a-zA-Z0-9_]*$/.test(name) 
        ) 
        {
          return false
        }
        return true
      }

    return (
        <div className="container">
            <h1>Create New Service</h1>
            <div role='form' className="addService-container">
                <div className="addService-titlebox" role = 'textbox' >
                    <span>name</span>
                    <input className = 'addName-input' pattern="[A-Za-z]{3}" onChange={(event) => {setName(event.target.value)}}/>
                </div>

                <div className="addService-descbox" role = 'textbox' >
                    <span>description</span>
                    <input onChange={(event) => {setDescription(event.target.value)}}/>
                </div>

                <div className="addService-imgbox">
                    <span>photo</span>
                    <label htmlFor="addService-file-Upload" className="addService-fileLabel">
                        {img? "" : "Upload Image"}
                        <div id="imgWrapper">
                            <img src={img? URL.createObjectURL(img) : null} alt={img? img.name : null} width="340" height="220"/>
                        </div>
                    </label>
                    <input id="addService-file-Upload" type="file" onChange={handleImage}/>
                </div>
            </div>
            <button className="actButton" onClick={() => handleSubmit()} disabled={validate()}>Create</button>
        </div>
    );
}

export default AddService;  