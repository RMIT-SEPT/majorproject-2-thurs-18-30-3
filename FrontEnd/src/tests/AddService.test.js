import React from 'react'

import AddService from '../components/AddService.js'
import {StaticRouter} from 'react-router'
import renderer from 'react-test-renderer'
import {render, screen, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom'
import '@testing-library/jest-dom/extend-expect'

const addFuncMock = jest.fn();

const setup = () => {
    return render(
      <StaticRouter>
        <AddService create={addFuncMock}/>
      </StaticRouter>
    )
}
  
  //snapshot test
  test('Test Page Renders', () => {
    const comp = renderer
      .create(
        <StaticRouter>
            <AddService create={addFuncMock}/>
        </StaticRouter>
      ).toJSON();
    
      expect(comp).toMatchSnapshot()
  })

  test('Test Page Contents', () => {
    const container = setup();
    expect(screen.getByRole('heading')).toHaveTextContent('Create New Service');
    expect(screen.getByRole('form')).toHaveClass('addService-container');
    expect(screen.getByRole('img')).toHaveAttribute('height','220');
    expect(screen.getByRole('img')).toHaveAttribute('width','340');
    expect(screen.getByRole('button')).toHaveTextContent('Create');
  })

  //Make sure create function isn't called when data is missing
  test('Test Validation', () => {
    const container = setup();
    const btn = screen.getByRole('button');
    const nameBox = screen.getAllByRole('textbox')[1];
    const descBox = screen.getAllByRole('textbox')[3];

    fireEvent.click(btn);
    expect(addFuncMock).toBeCalledTimes(0);

    fireEvent.change(nameBox, {target: {value: 'test'}})
    fireEvent.click(btn);
    expect(addFuncMock).toBeCalledTimes(0);

    fireEvent.change(nameBox, {target: {value: ''}})
    fireEvent.change(descBox, {target: {value: 'test'}})
    fireEvent.click(btn);
    expect(addFuncMock).toBeCalledTimes(0);
  })

  //Make sure create function is called correctly
  test('Test Confirms', () => {
    const container = setup();
    const btn = screen.getByRole('button');
    const nameBox = screen.getAllByRole('textbox')[1];
    const descBox = screen.getAllByRole('textbox')[3];

    fireEvent.change(nameBox, {target: {value: 'test name'}})
    fireEvent.change(descBox, {target: {value: 'test desc'}})
    fireEvent.click(btn);
    expect(addFuncMock).toBeCalled();
  })