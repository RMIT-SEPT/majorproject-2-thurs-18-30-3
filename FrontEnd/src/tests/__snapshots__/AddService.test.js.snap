// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Page Renders 1`] = `
<div
  className="container"
>
  <h1>
    Create New Service
  </h1>
  <div
    className="addService-container"
    role="form"
  >
    <div
      className="addService-titlebox"
      role="textbox"
    >
      <span>
        name
      </span>
      <input
        className="addName-input"
        onChange={[Function]}
        pattern="[A-Za-z]{3}"
      />
    </div>
    <div
      className="addService-descbox"
      role="textbox"
    >
      <span>
        description
      </span>
      <input
        onChange={[Function]}
      />
    </div>
    <div
      className="addService-imgbox"
    >
      <span>
        photo
      </span>
      <label
        className="addService-fileLabel"
        htmlFor="addService-file-Upload"
      >
        Upload Image
        <div
          id="imgWrapper"
        >
          <img
            alt={null}
            height="220"
            src={null}
            width="340"
          />
        </div>
      </label>
      <input
        id="addService-file-Upload"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <button
    className="actButton"
    disabled={true}
    onClick={[Function]}
  >
    Create
  </button>
</div>
`;
